core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java

```diff
@@ -926,30 +999,57 @@ public Command workerHeartbeat(long workerId, Map<String, Long> capacityBytesOnT
       return Command.newBuilder().setCommandType(CommandType.Register).build();
     }
 
-    synchronized (worker) {
-      // Technically, 'worker' should be confirmed to still be in the data structure. Lost worker
-      // detection can remove it. However, we are intentionally ignoring this race, since the worker
-      // will just re-register regardless.
-      processWorkerRemovedBlocks(worker, removedBlockIds);
-      processWorkerAddedBlocks(worker, addedBlocks);
-      processWorkerMetrics(worker.getWorkerAddress().getHost(), metrics);
+    // Update the TS before the heartbeat so even if the worker heartbeat processing
+    // is time-consuming or triggers GC, the worker does not get marked as lost
+    // by the LostWorkerDetectionHeartbeatExecutor.
+    worker.updateLastUpdatedTimeMs();
 
+    // The address is final, no need for locking
+    processWorkerMetrics(worker.getWorkerAddress().getHost(), metrics);
+
+    worker.getUsageLock().writeLock().lock();
+    worker.getBlockListLock().writeLock().lock();
+    Command workerCommand = null;
+    try {
       worker.addLostStorage(lostStorage);
 
       if (capacityBytesOnTiers != null) {
         worker.updateCapacityBytes(capacityBytesOnTiers);
       }
       worker.updateUsedBytes(usedBytesOnTiers);
-      worker.updateLastUpdatedTimeMs();
 
+      // Technically, 'worker' should be confirmed to still be in the data structure. Lost worker
+      // detection can remove it. However, we are intentionally ignoring this race, since the worker
+      // will just re-register regardless.
+
+      // This changes the block list
+      processWorkerRemovedBlocks(worker, removedBlockIds);
+      processWorkerAddedBlocks(worker, addedBlocks);
+      // TODO(jiacheng): If the worker registers with a after the block is freed,
+      //  the copy will not be removed from this worker.
       List<Long> toRemoveBlocks = worker.getToRemoveBlocks();
       Metrics.TOTAL_BLOCKS.inc(addedBlocks.size() - removedBlockIds.size());
       if (toRemoveBlocks.isEmpty()) {
-        return Command.newBuilder().setCommandType(CommandType.Nothing).build();
+        workerCommand = Command.newBuilder().setCommandType(CommandType.Nothing).build();
+      } else {
+        workerCommand = Command.newBuilder().setCommandType(CommandType.Free)
+            .addAllData(toRemoveBlocks).build();
       }
-      return Command.newBuilder().setCommandType(CommandType.Free).addAllData(toRemoveBlocks)
-          .build();
+    } finally {
+      // Release in the reverse order of acquisition
+      worker.getBlockListLock().writeLock().unlock();
+      worker.getUsageLock().writeLock().unlock();
     }
+
+    // Update the TS again
+    worker.updateLastUpdatedTimeMs();
+
+    // Should not reach here
+    if (workerCommand == null) {
+      LOG.error("Worker heartbeat response command is null!");
```

Comment:
Better way to handle this like `Preconditions.checkNotNull`?

Reply:
If the code should not reach here, then why not throw with `Preconditions.checkNotNull`?
It's more like `Postconditions` though :/

Reply:
Yea exactly why i didn't put a `Preconditions` here :)